bazel_tools/dev_env_tool/dev_env_tool.bzl

```diff
@@ -0,0 +1,140 @@
+# Copyright (c) 2019 The DAML Authors. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+load("@bazel_tools//tools/cpp:lib_cc_configure.bzl", "get_cpu_value")
+
+def _create_build_content(rule_name, tools, win_paths, nix_paths):
+    content = """
+# DO NOT EDIT: automatically generated BUILD file for dev_env_tool.bzl: {rule_name}
+package(default_visibility = ["//visibility:public"])
+
+filegroup(
+    name = "all",
+    srcs = glob(["**"]),
+)
+        """.format(rule_name = rule_name)
+
+    for i in range(0, len(tools)):
+        content += """
+filegroup(
+    name = "{tool}",
+    srcs = select({{
+        ":windows": ["{win_path}"],
+        "//conditions:default": ["{nix_path}"],
+    }}),
+)
+            """.format(
+            tool = tools[i],
+            win_path = win_paths[i],
+            nix_path = nix_paths[i],
+        )
+
+    content += """
+config_setting(
+    name = "windows",
+    values = {"cpu": "x64_windows"},
+    visibility = ["//visibility:private"],
+)
+"""
+
+    return content
+
+def _dadew_where(ctx, ps):
+    ps = ctx.which("powershell")
+    ps_result = ctx.execute([ps, "-Command", "dadew enable; dadew where"], quiet = True)
+
+    if ps_result.return_code != 0:
+        fail("Failed to obtain dadew location.\nExit code %d.\n%s\n%s" %
+             (ps_result.return_code, ps_result.stdout, ps_result.stderr))
+
+    return ps_result.stdout.splitlines()[0]
+
+def _dadew_tool_home(dadew, tool):
+    return "%s\\scoop\\apps\%s\\current" % (dadew, tool)
+
+def _find_files_recursive(ctx, find, root):
+    find_result = ctx.execute([find, "-L", root, "-type", "f", "-printf", "%P\\0"])
+
+    if find_result.return_code != 0:
+        fail("Failed to list files contained in '%s':\nExit code %d\n%s\n%s." %
+             (root, find_result.return_code, find_result.stdout, find_result.stderr))
+
+    return [f for f in find_result.stdout.split("\0") if f]
+
+def _symlink_files_recursive(ctx, find, source, dest):
+    files = _find_files_recursive(ctx, find, source)
+    for f in files:
+        ctx.symlink("%s/%s" % (source, f), "%s/%s" % (dest, f))
+    if not files:
+        # If no files where found source might be an empty directory or a file.
```

Comment:
I don’t understand how we can get here if `source` is a file. `find -type f file` will return the file afaik so it should be nonempty in this case. What am I missing?

Reply:
That's a side-effect of `-printf "%P"`. This tells `find` to print the path relative to the starting-point. If the starting point is a file, then it prints no output.

However, turns out that BSD find does not support `-printf` and we have to do manual string mangling to get relative paths anyway, so this is gone now.

Reply:
Thanks for the explanation!

Reply:
You could maybe run `find .` with the current working directory as `root`. This should make it easier.

```python
>>> root = os.path.join(os.getcwd(), 'bazel_tools')
>>> root
'/Users/samirtalwar/src/digital-asset/daml/bazel_tools'
>>> found = subprocess.run(['find', '.', '-type', 'd'], cwd=root, stdout=subprocess.PIPE, encoding='utf-8').stdout.strip().split('\n')
>>> found
['.', './packaging', './client_server_test', './client_server_test/runner', './client_server_test/tests', './docs', './runfiles', './runfiles/src', './runfiles/src/main', './runfiles/src/main/scala', './runfiles/src/main/scala/com', './runfiles/src/main/scala/com/digitalasset', './runfiles/src/main/scala/com/digitalasset/daml', './runfiles/src/main/scala/com/digitalasset/daml/bazeltools', './java_testing', './dev_env_package']
>>> [os.path.normpath(d) for d in found]
['.', 'packaging', 'client_server_test', 'client_server_test/runner', 'client_server_test/tests', 'docs', 'runfiles', 'runfiles/src', 'runfiles/src/main', 'runfiles/src/main/scala', 'runfiles/src/main/scala/com', 'runfiles/src/main/scala/com/digitalasset', 'runfiles/src/main/scala/com/digitalasset/daml', 'runfiles/src/main/scala/com/digitalasset/daml/bazeltools', 'java_testing', 'dev_env_package']

Reply:
I guess you can't use `os.path` in Bazel (though I have no real clue) but hopefully it'll still make it cleaner?

Reply:
@SamirTalwar Yes, that would be a nice simplification. Sadly, I haven't found a way to make it work in Bazel. [`repository_ctx.execute`](https://docs.bazel.build/versions/master/skylark/lib/repository_ctx.html#execute) does allow to change the working directory. But, if `root` is a file, which is e.g. the case with `@tar_dev_env`, then this will fail. The `repository_ctx` API does not provide an easy way to distinguish between files and directories, so we can't just branch.

Reply:
Aha, I didn't expect that. Thanks for clarifying. 😄